### PR TITLE
Compiler plugin for implementing incremental computation in user space

### PR DESCRIPTION
**Pull request by cooldome**
This plugin provides essential building block for implementing `incremental computations` in your programs. The idea behind incremental computations is that if you do the same calculation multiple times but with slightly different inputs you don't have to recompute everything from scratch. Also you don't want to adopt special algorithms either, you would like to write your code in standard from scratch manner and get _incrementality_ for free when it is possible.

The plugin computes the digest of the proc bodies, recursively hashing all called procs as well . Such digest with the digest of the argument values gives a good "name" for the result. Terminology loosely follows paper "Incremental Computation with Names" link below. It works well if you have no side effects in your computations. If you have global state in your computations then you will need problem specific workarounds to represent global state in set of "names" . SideEffect tracking in Nim also useful in this topic.

Classical examples: 
  * Dashboard with ticking data. New data arrives non stop and you would like to update the dashboard recomputing only changed outputs.
  * Excel spreadsheet where user changes one cell and you would like to recompute all cells that are affected by the change, but do not want to recompute every cell in the spreadsheet.

Good tutorials in this topic:

- http://en.wikipedia.org/wiki/Incremental_computing
- https://www.youtube.com/watch?v=G6a5G5i4gQU
- https://vimeo.com/122066659

Papers:

- https://arxiv.org/abs/1503.07792
- https://www.cs.cmu.edu/~rwh/theses/acar.pdf
- 

Websites:

- http://adapton.org/
- http://www.umut-acar.org/self-adjusting-computation

Frequently Asked Question:
Q: Is compiler plugin necessary here, can't it be implemented in the user space with the VM?
A: It is possible to implement without plugin with VM only and it was done. What was found that it takes 10 sec to compute one digest in the VM, but took only 2 ms as compiler plugins. Hence the plugin.

Q: Is it really a compiler plugin?
A: Not quite. Nim has infrastructure for compiler plugins, but it is for semantic pass only and such plugins don't work in VM. VM support here is essential, hence extra VM op is provided instead.
If you find it confusing I can move the code to the `sighashes.nim` instead and don't call it a plugin.

Q: Nim compiler already has hashing for procs, how it is different from existing proc hashing?
A: The existing hashing takes into account types and owners, while this one hashes also bodies
and called procs as well. Global cache of proc digests is used as the same procs are used over and over again.

Q: Is this plugin related to the compiler's incremental compilation?
A: Not really, but it might be useful in this context as well. 

Q. PR also includes `readLines` and `getMD5` in VM what is it?
A: It helps not only to incrementalize the runtime, but also the macros codegen of files. We are computing the digest of the typed proc and macro and compare with digest saved in the header of the file.

### CODE REVIEW

**Comment by andreaferretti** [comment]
Really cool, but is there a test showing how this is used?

**Comment by cooldome** [comment]
Agree, I will add extra test with example

**Comment by cooldome** [comment]
I have added the example test case. The test case is simplistic and in real world you will have to do a lot more: use dedicated cache containers, handle arguments, store/load to disk and etc.

It makes the following example works:
```nim
func input1(): float {.incremental_input("a1").}

func input2(): float {.incremental_input("a2").}

func sub_calc1(a: float): float  {.incremental.} = 
  a + input1()

func sub_calc2(b: float): float  {.incremental.} = 
  b + input2()

func heavy_calc(a: float, b: float): float {.incremental.} = 
  sub_calc1(a) + sub_calc2(b)

######################################


set_input("a1", 5)
set_input("a2", 2)
discard heavy_calc(5.0, 10.0)

echo "** change one input and recompute effectively"

set_input("a2", 10)
discard heavy_calc(5.0, 10.0) # only one branch of the graph should be recomputed
```
The names used are stable across the builds with code changes reflected automatically in the names. It allow one to persist the cache and reuse later even if the code has partially changed.

Ready for review.

File: compiler/plugins/symdigest.nim [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    return value[]
+
+  var c: MD5Context
+  md5Init(c)
+  c.hashType(sym.typ, {CoProc})
+  c &= sym.kind
+  c.md5Final(result)
+  procSymHashes[sym.id] = result # protect from recursion in the body
+
+  if sym.ast != nil and sym.ast[bodyPos] != nil:
```
Comment (Araq) [inline compiler/plugins/symdigest.nim]: ``sym.ast[bodyPos]`` can never be nil by design. For defensive coding you could use ``bodyPos < sym.ast.safeLen``.

File: compiler/plugins/symdigest.nim [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    c &= n.intVal
+  of nkFloatLit..nkFloat64Lit:
+    c &= n.floatVal
+  of nkStrLit..nkTripleStrLit:
+    c &= n.strVal
+  else:
+    for i in 0..<n.len:
+      hashTree(c, n.sons[i])
+
+var procSymHashes {.global.} = initTable[int, MD5Digest](2)
```
Comment (Araq) [inline compiler/plugins/symdigest.nim]: I don't like global state in the compiler, the compiler is also a library. You need to pass this around everywhere in some context. And then attach the context to ``semdata.PContext`` or similar.

File: compiler/vmdef.nim [hunk 3]
```diff
@@ -93,6 +93,7 @@ type
     opcNGetType,
     opcNStrVal,
     opcNSigHash,
+    opcNSymBodyHash
```
Comment (Araq) [inline compiler/vmdef.nim]: Instead of yet another VM opcode, can you instead use the VM's callback mechanism?

File: lib/system/io.nim [hunk 4]
```diff
@@ -612,6 +612,25 @@ proc writeFile*(filename, content: string) {.tags: [WriteIOEffect], benign.} =
   else:
     sysFatal(IOError, "cannot open: " & filename)
 
+
+proc readLines*(filename: string, n = 1.Natural): seq[TaintedString] =
```
Comment (Araq) [inline lib/system/io.nim]: Please don't add this to ``io.nim``, there is got to be a better way.

**Comment by zah** [comment]
I this the "plugins" folder is intended for modules that are optional and can be enabled on-demand during compilation, so this can be moved in sighashes as suggested.

File: compiler/plugins/symdigest.nim (hunk 2 above)
Comment (cooldome) [reply compiler/plugins/symdigest.nim]: As far I got it `semdata.PContext` is per module while this cache is cross module. is it ok if I will attached to `ModuleGraph`?

File: lib/system/io.nim (hunk 4 above)
Comment (cooldome) [reply lib/system/io.nim]: I thought what can be a better way. The only idea that might be questionally better, is to make this example work in compile time:
```nim
static:
  var file = open(file, "copying.txt"):
  for s in lines(file):
    echo s
  close(file)
```
with some hackery in vmops.nim (ptr File <-> IntLitNode casting) and vm callbacks I think it is possible. Is this something you would prefer? the only other option I can think is a separate `readLines` proc just like in this PR.

**Comment by cooldome** [comment]
I will move code to sighashes

File: lib/system/io.nim (hunk 4 above)
Comment (Araq) [reply lib/system/io.nim]: Well if it's only used by the test case, move it to the test case.

Comment (cooldome) [reply lib/system/io.nim]: Introduced `readLines` is used at compile time and there is also vm callback added for this proc. 
Comment mentions that it is available at compile time.
It is used to read the `discard "asasdf687687a6sdf7"` header of the file to check if the file needs to be regenerated by nim macro.

File: compiler/plugins/symdigest.nim (hunk 2 above)
Comment (Araq) [reply compiler/plugins/symdigest.nim]: Yes.

**Comment by cooldome** [comment]
I all review comments are now resolved. Moving hashing code into sighashes introduced circular dependency between ModuleGraph and SigHash types, therefore I had to move SigHash type from sighashes.nim to modulegraphs.nim. Not ideal, but in my opinion is still better than separate file just for types. Let me know of you prefer different location for SigHash type.

Ready for review.

File: compiler/plugins/symdigest.nim (hunk 1 above)
Comment (cooldome) [reply compiler/plugins/symdigest.nim]: done

File: compiler/vmdef.nim (hunk 3 above)
Comment (cooldome) [reply compiler/vmdef.nim]: done,  callback mechanism worked well

File: lib/core/macros.nim [hunk 5]
```diff
@@ -360,6 +360,13 @@ when defined(nimHasSignatureHashInMacro):
     ## the owning module of the symbol and others. The same identifier is
     ## used in the back-end to produce the mangled symbol name.
 
+when defined(nimHasSymBodyHashInMacro):
+  proc symBodyHash*(s: NimNode): string {.magic: "NSymBodyHash", noSideEffect.}
```
Comment (Araq) [inline lib/core/macros.nim]: Since this is now a VM callback there is no need to make this a ``magic``.

Comment (cooldome) [reply lib/core/macros.nim]: sure and defined(nimHasSymBodyHashInMacro) is also no longer required.
All removed looks better now.

**Comment by cooldome** [comment]
ready for review

**Comment by cooldome** [comment]
minor correction

**Comment by cooldome** [comment]
ready for review

**Comment by Araq** [comment]
Sorry this took so long.

**Pull request merged**